compiler_opt/rl/corpus.py

```diff
@@ -30,9 +33,74 @@ class ModuleSpec:
   """
   name: str
   exec_cmd: Tuple[str, ...] = ()
+  size: int = 0
 
 
-def build_modulespecs_from_datapath(
+class Corpus:
+  """Represents a corpus. Comes along with some utility functions."""
+
+  def __init__(self,
+               data_path: str,
+               additional_flags: Tuple[str, ...] = (),
+               delete_flags: Tuple[str, ...] = ()):
+    self._module_specs = _build_modulespecs_from_datapath(
+        data_path=data_path,
+        additional_flags=additional_flags,
+        delete_flags=delete_flags)
+    self.root_dir = data_path
+    self._module_specs.sort(key=lambda m: m.size, reverse=True)
+
+  @classmethod
+  def from_module_specs(cls, module_specs):
+    """Construct a Corpus from module specs. Mostly for testing purposes."""
+    corp = cls.__new__(cls)  # Avoid calling __init__
+    super(cls, corp).__init__()
+    corp._module_specs = list(module_specs)  # Don't mutate the original list.
+    corp._module_specs.sort(key=lambda m: m.size, reverse=True)
+    corp.root_dir = None
+    return corp
+
+  def sample(self,
+             k: int,
+             sort: bool = False,
+             sampler=random.sample) -> List[ModuleSpec]:
+    """Samples `k` module_specs, sorting by size descending."""
+    k = min(len(self._module_specs), k)
+    if k < 1:
+      raise ValueError('Attempting to sample <1 module specs from corpus.')
+    sampled_specs = sampler(self._module_specs, k=k)
+    if not sort:
+      return sampled_specs
+    return list(sorted(sampled_specs, key=lambda m: m.size, reverse=True))
+
+  def filter(self, p: re.Pattern):
+    """Filters module specs, keeping those which match the provided pattern."""
+    self._module_specs = [ms for ms in self._module_specs if p.match(ms.name)]
+
+  def __len__(self):
+    return len(self._module_specs)
+
+
+def sampler_bucket_round_robin(module_specs: List[ModuleSpec],
```

Comment:
the current logic is quite complicated, why not directly do sampling?

Reply:
i get a bit confused of the function, originally i thought you want to sample k modules that are of similar size so that the time to compile them was similar so that the data collection can be more efficient

it seems that what you are doing here is something reverse --- have a more spread distribution of compilation time, i'm wondering what's the benefit of it compared with just a random sampling?